pyccel/parser/semantic.py

```diff
@@ -3659,88 +3659,123 @@ def _visit_For(self, expr):
+                var_name = self.scope.get_expected_name(expr.indices[i])
+                handle_int_loop_variable(var_name)
                 start = a.start
+                stop  = a.stop
                 step  = a.step
 
-            elif isinstance(a, (PythonZip, PythonEnumerate)):
-                dvar  = self._infer_type(a.element)
-                class_type = dvar.pop('class_type')
-                if class_type.rank > 0:
-                    class_type = class_type.switch_rank(class_type.rank-1)
-                    dvar['shape'] = (dvar['shape'])[1:]
-                if class_type.rank == 0:
-                    dvar['shape'] = None
-                    dvar['memory_handling'] = 'stack'
-                var  = Variable(class_type, var, **dvar)
+            elif isinstance(a, PythonEnumerate):
+                var_name1 = self.scope.get_expected_name(expr.indices[i][0])
+                var_name2 = self.scope.get_expected_name(expr.indices[i][1])
+                handle_int_loop_variable(var_name1)
+                handle_iterable_variable(var_name2, a.element)
                 stop = a.element.shape[0]
 
-            elif isinstance(a, Variable):
-                dvar  = self._infer_type(a)
-                class_type = dvar.pop('class_type')
-                if class_type.rank > 0:
-                    class_type = class_type.switch_rank(class_type.rank-1)
-                    dvar['shape'] = (dvar['shape'])[1:]
-                if class_type.rank == 0:
-                    dvar['shape'] = None
-                    dvar['memory_handling'] = 'stack'
-                var  = Variable(class_type, var, **dvar)
-                stop = a.shape[0]
+            elif isinstance(a, PythonZip):
+                for idx, arg in enumerate(a.args):
+                    var = self.scope.get_expected_name(expr.indices[i][idx])
+                    handle_iterable_variable(var, arg)
+                stop = a.get_range().stop
+
+            elif isinstance(a, VariableIterator):
+                var = self.scope.get_expected_name(expr.indices[i])
+                handle_iterable_variable(var, a.variable)
+                stop = a.variable.shape[0]
 
             else:
                 errors.report(PYCCEL_RESTRICTION_TODO,
                               bounding_box=(self.current_ast_node.lineno, self.current_ast_node.col_offset),
                               severity='fatal')
-            existing_var = self.scope.find(var.name, 'variables')
-            if existing_var:
-                if self._infer_type(existing_var) != dvar:
-                    errors.report(f"Variable {var} already exists with different type",
+
+            for var, dvar in variables:
+                existing_var = self.scope.find(var.name, 'variables')
+                if var.name == expr.lhs:
+                    return errors.report(f"Variable {var} has the same name as the left hand side",
                             symbol = expr, severity='error')
```

Comment:
Do you have an example of where this can arise?

Reply:
Yes.
```python
a = [a for a in range(10)]
```
Could you please point out where to add tests that should raise errors so I can add this kind of examples?

Reply:
https://github.com/pyccel/pyccel/tree/devel/tests/errors